pkg/executor/insert.go

```diff
@@ -236,14 +237,42 @@ func (e *InsertExec) batchUpdateDupRows(ctx context.Context, newRows [][]types.D
 		e.stats.Prefetch += time.Since(prefetchStart)
 	}
 
+	// If the current row has some conflicts, the operation will be changed to update.
+	// If this happens, there still may be another index that has a conflict,
+	// so we need to determine DupKeyCheckMode here.
+	updateDupKeyCheck := table.DupKeyCheckInPlace
+	if (txn.IsPessimistic() && !e.IgnoreErr) || txn.IsPipelined() {
+		// - If `txn.Pipelined()`, it means current is using `@@tidb_dml_type="bulk"` to insert rows.
+		//   `DupKeyCheckLazy` should be used in "bulk" mode to avoid request storage and improve the performance.
+		// - If `txn.IsPessimistic()`, we can use `DupKeyCheckLazy` to postpone the storage constraints check
+		//   to subsequence stages such as lock.
+		//   However, if the current statement is `INSERT IGNORE ... ON DUPLICATE KEY ...`,
+		//   `DupKeyCheckInPlace` should be used.
+		//   It is because the executor should get the dup-key error immediately and ignore it.
+		// - If the current txn is optimistic, `DupKeyCheckInPlace` is always used
+		//   even if `tidb_constraint_check_in_place` is `OFF`.
+		//   This is because `tidb_constraint_check_in_place` is only designed for insert cases, see comments in issue:
+		//   https://github.com/pingcap/tidb/issues/54492#issuecomment-2229941881
+		//   Though it is still in an insert statement, but it seems some old tests still think it should
+		//   check constraints in place, see test:
+		//   https://github.com/pingcap/tidb/blob/3117d3fae50bbb5dabcde7b9589f92bfbbda5dc6/pkg/executor/test/writetest/write_test.go#L419-L426
```

Comment:
`tidb_constraint_check_in_place_pessimistic` is not considerate yet. It will be handled in another PR to introduce a new option like `PessimisticLazyCheckMode` which only works for pessimistic txn and `DupKeyCheckMode` is `DupKeyCheckLazy`